add doc for v5.1: Deploy -> Config TiKV (#233)
tikv/website: content/docs/5.1/deploy/configure/security.md

---
title: Security

> description: Keeping your TiKV deployment secure

Reviewer: ```suggestion
description: Keeping your TiKV secure
```

menu:
    "5.1":
        parent: Configure TiKV
        weight: 5
---

> This page discusses how to use [Transport Layer Security](#transport-layer-security-tls) to encrypt connections between TiKV nodes.

Reviewer: ```suggestion
This section describes how to use [Transport Layer Security(TLS)](#transport-layer-security-tls) to encrypt connections between TiKV nodes.
```

> ## Transport Layer Security (TLS)

Reviewer: ```suggestion
## Transport Layer Security
```

```bash
#! /bin/bash
set +e

mkdir -p easyrsa
cd easyrsa
curl -L https://github.com/OpenVPN/easy-rsa/releases/download/v3.0.6/EasyRSA-unix-v3.0.6.tgz \
    | tar xzv --strip-components=1

./easyrsa init-pki \
    && ./easyrsa build-ca nopass

NUM_PD_NODES=3
for i in $(seq 1 $NUM_PD_NODES); do
    ./easyrsa gen-req pd$i nopass
    ./easyrsa sign-req server pd$i
done

NUM_TIKV_NODES=3
for i in $(seq 1 $NUM_TIKV_NODES); do
    ./easyrsa gen-req tikv$i nopass
    ./easyrsa sign-req server tikv$i
done

./easyrsa gen-req client nopass
./easyrsa sign-req server client
```

If you run this script, you'll need to interactively answer some questions and make some confirmations. You can answer with anything for the CA common name. For the PD and TiKV nodes, use the hostnames.

If the script runs successfully, you should have something like this:

```bash
$ ls easyrsa/pki/{ca.crt,issued,private}
easyrsa/pki/ca.crt

easyrsa/pki/issued:
client.crt  pd1.crt  pd2.crt  pd3.crt  tikv1.crt  tikv2.crt  tikv3.crt

easyrsa/pki/private:
ca.key  client.key  pd1.key  pd2.key  pd3.key  tikv1.key  tikv2.key  tikv3.key
```

### Configure the TiKV Server Certificates

> Specify the TLS options for TiKV certificate with the configuration file options:

Reviewer: ```suggestion
Set the certificate in TiKV configuration file:
```

```toml
# Using empty strings here means disabling secure connections.
[security]
# The path to the file that contains the PEM encoding of the server’s CA certificates.
ca-path = "/path/to/ca.pem"
# The path to the file that contains the PEM encoding of the server’s certificate chain.
cert-path = "/path/to/tikv-server-cert.pem"
# The path to the file that contains the PEM encoding of the server’s private key.
key-path = "/path/to/tikv-server-key.pem"
# The name list used to verify the common name in client’s certificates. Verification is
# not enabled if this field is empty.
cert-allowed-cn = ["tikv-server", "pd-server"]
```

You'll also need to **change the connection URL to `https://`**.

You can find all the TiKV TLS configuration options [here](../tikv-configuration-file/#security).

### Configure the PD Certificates

> Specify the TLS options for PD certificate with the configuration file options:

Reviewer: ```suggestion
Set the certificate in PD configuration file:
```

```toml
[security]
# The path to the file that contains the PEM encoding of the server’s CA certificates.
cacert-path = "/path/to/ca.pem"
# The path to the file that contains the PEM encoding of the server’s certificate chain.
cert-path = "/path/to/pd-server-cert.pem"
# The path to the file that contains the PEM encoding of the server’s private key.
key-path = "/path/to/pd-server-key.pem"
# The name list used to verify the common name in client’s certificates. Verification is
# not enabled if this field is empty.
cert-allowed-cn = ["tikv-server", "pd-server"]
```

You'll also need to **change the connection URL to `https://`**.

You can find all the PD TLS configuration options [here](../pd-configuration-file/#security).

### Configure the Client

> When connecting your TiKV Client, you'll need to specify the TLS options. In this example, we build a configuration for the [Rust Client](https://github.com/tikv/client-rust):

Reviewer: ```suggestion
You need to set TLS options for TiKV Client in order to connect to the TiKV. Taking [Rust Client](https://github.com/tikv/client-rust) as example, TLS option is set like this:
```

```rust
let config = Config::new(/* ... */).with_security(
    // The path to the file that contains the PEM encoding of the server’s CA certificates.
    "/path/to/ca.pem",
    // The path to the file that contains the PEM encoding of the server’s certificate chain.
    "/path/to/client-cert.pem",
    // The path to the file that contains the PEM encoding of the server’s private key.
    "/path/to/client-key.pem"
);
```

> You'll also need to **change the connection URL to `https://`**.

Reviewer: ```suggestion
You'll also need to **change the connection URL to `https://`** instead of plain `ip:port`.
```

{{< warning >}}
Currently TiKV Java Client does not support TLS.
{{< /warning >}}

### Connecting with `tikv-ctl` and `pd-ctl`

> When using `pd-ctl` and `tikv-ctl` the relevant options will need to be specified:

Reviewer: ```suggestion
When using `pd-ctl` and `tikv-ctl` the relevant options will need to be set:
```
